### PR TITLE
aggregatekeylinks subcommand

### PR DESCRIPTION
**Pull request by yilmazbahadir**
- aggregatekeylinks subcommand to aggregate accountKeyLink,votingKeyLink,vrfKeyLink txs and optional the enrol tx
- invalid aggregate tx fix: there has to be an inner tx signed by the agg. bonded tx signer

### CODE REVIEW

**Comment by segfaultxavi** [comment]
I sent the agg tx from symbol-cli without issues. Then I signed from the Waller.

Signing with the Wallet had this UI issue:
image

But I was able to sign! Good progress!

However, I got a `Failure_Core_Link_Already_Exists` error:
http://explorer.testnet.symboldev.network/transactions/9748AA28CAA75D31E89EA2C5588058207F30E4CFEF288C057366E56544C191C4
The account is not linked to anything, as far as I can see:
http://explorer.testnet.symboldev.network/accounts/TAZBGUACETXZHAQEQZACFHTLE7OYG2FMGJHZ2SQ

**Comment by segfaultxavi** [comment]
I get the same error (`Failure_Core_Link_Already_Exists`) when signing from symbol-cli.

**Comment by segfaultxavi** [comment]
I also get the same error with a brand new customer account, so there's no chance it is already linked to anything.

**Comment by yilmazbahadir** [comment]
TAZBGUACETXZHAQEQZACFHTLE7OYG2FMGJHZ2SQ is the customer's public key right?
Can you give me the signer account's public key? Maybe it is trying to link with it?

**Comment by segfaultxavi** [comment]
The signer's pubkey is `5444F0637CE592F595C76A7B431652506763442515C766FBE9A22E28DFB08FA7`.

**Comment by yilmazbahadir** [comment]
Can you try setting  these parameters with the exact values? => `--voting-start-point 400 --voting-end-point 1000 `
I tried and it worked for me. Because I found the code piece in the image below in the catapult-server that checks the starting epoch(voting-start-point) and give that error, setting it to 400 should do it for now.

Can you please find out how to set those values and what do they actually mean and document if possible please?

image

**Comment by segfaultxavi** [comment]
YES!!! That did it! It worked for me

We need @Jaguar0625 to tell us what's the purpose of that line and if that's the right error message to throw there. It would have helped a lot if the error pointed somehow to the start epoch.

What about the UI problem, @yilmazbahadir, do you see it?

File: src/commands/transaction/aggregatekeylinks.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    })
+    remoteLinkAction: string;
+
+    @option({
+        description: 'Linked Voting Account Public Key.',
+    })
+    votingPublicKey: string;
+
+    @option({
+        description: 'Voting Key Link Start Point.',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:51]: Point or Epoch?

File: src/commands/transaction/aggregatekeylinks.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    })
+    votingStartPoint: string;
+
+    @option({
+        description: 'Voting Key Link End Point.',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:56]: Point or Epoch?

File: src/commands/transaction/aggregatekeylinks.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    })
+    votingEndPoint: string;
+
+    @option({
+        description: 'Voting Key Link action (Link, Unlink).',
+    })
+    votingLinkAction: string;
+
+    @option({
+        description: 'Linked Vrf Account Public Key.',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:66]:
```suggestion
        description: 'Linked VRF Account Public Key.',
```

File: src/commands/transaction/aggregatekeylinks.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    })
+    vrfPublicKey: string;
+
+    @option({
+        description: 'Vrf Key Link action (Link, Unlink).',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:71]:
```suggestion
        description: 'VRF Key Link action (Link, Unlink).',
```

File: src/commands/transaction/aggregatekeylinks.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    })
+    vrfLinkAction: string;
+
+    @option({
+        flag: 't',
+        description: 'Aggregate Type (complete or bonded)',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:77]:
```suggestion
        description: 'Aggregate Type (AGGREGATE_BONDED or AGGREGATE_COMPLETE)',
```

File: src/commands/transaction/aggregatekeylinks.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    @option({
+        description: '(Optional) Enrol Recipient Address',
+    })
+    enrolRecipientAdress: string;
+}
+
+@command({
+    description:
+        'Aggregates all key link transactions(Remote Account, Voting, Vrf) and optionally enrol transaction needed to set up a node.',
```
Comment (segfaultxavi) [inline src/commands/transaction/aggregatekeylinks.ts:104]:
```suggestion
        'Wraps all key link transactions(Remote Account, Voting, VRF) and optionally a SuperNode enrol transaction with an aggregate bonded transaction to be cosigned by another account. This is useful to setup a node from one account but keep the main account (the one with the funds) isolated.',
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
Polished texts a bit.

**Comment by segfaultxavi** [comment]
This command only makes sense for aggregate BONDED transactions, right? I don't think we need to ask for the Aggregate Type.
Moreover, if you try to use the COMPLETE type, you get:
```
TypeError: Cannot read property 'minApproval' of undefined
    at TransactionSignatureService.get announceMode [as announceMode] (/home/xartigas/repos/symbol-cli/build/src/services/transaction.signature.service.js:47:88)
    at TransactionSignatureService.signTransactions (/home/xartigas/repos/symbol-cli/build/src/services/transaction.signature.service.js:81:18)
    at default_1.signTransactions (/home/xartigas/repos/symbol-cli/build/src/interfaces/announce.transactions.command.js:82:101)
    at default_1.execute (/home/xartigas/repos/symbol-cli/build/src/commands/transaction/aggregatekeylinks.js:163:47)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
```

If we want to allow COMPLETE transactions (maybe to simplify node setup?) then we do not need to ask for the cosigner.

File: src/commands/transaction/aggregatekeylinks.ts (hunk 2 above)
Comment (yilmazbahadir) [reply src/commands/transaction/aggregatekeylinks.ts:56]: It is stated as Point in the votingKeyLink subcommand
https://github.com/nemtech/symbol-cli/blob/main/src/commands/transaction/votingkeylink.ts

**Comment by yilmazbahadir** [comment]
There are several use cases we can support :

1.  Acc1 prepares the AGGREGATE_BONDED, adds the inner transactions for Acc2 to sign **online**, Acc1 announces the tx, Acc2 cosigns in the network(online).
2. Acc1 prepares the AGGREGATE_COMPLETE, adds the inner transactions for Acc2 to sign **offline**, Acc1 doesn't announce the tx instead sends the PAYLOAD to the Acc2 to sign offline and then Acc1 announces to the network
3. Acc2 prepares the AGGREGATE_COMPLETE, with the inner transactions and announces to the network.

What do you think, should we support all 3 of them or support only the option 1 ?

File: src/commands/transaction/aggregatekeylinks.ts (hunk 2 above)
Comment (segfaultxavi) [reply src/commands/transaction/aggregatekeylinks.ts:56]: OK, I don't understand yet what either means anyway.

**Comment by segfaultxavi** [comment]
Yeah, OK, all 3 use cases makes sense. Let's keep the Aggregate Complete option.

**Comment by segfaultxavi** [comment]
Everything looks fine to me now... except the command name, I'm afraid it's too generic, and the transfer tx is not actually a link.
It should convey the meaning of node setup somehow. So maybe `nodesetup`, `quicknodesetup`, `fullnodesetup`, `setupnodelinks`...

**Comment by yilmazbahadir** [comment]
Okay, pick one.

**Comment by segfaultxavi** [comment]
Let's use `setupnodelinks`.

**Review by segfaultxavi — APPROVED** [review]
I approve the text, somebody else should review the code...

**Review by rg911 — APPROVED** [review]

**Pull request merged**